core/arch/riscv/mm/core_mmu_arch.c

```diff
@@ -0,0 +1,653 @@
+					(pa_to_ppn((paddr_t)pgt)));
+	}
+
+	return true;
+}
+
+void core_mmu_set_info_table(struct core_mmu_table_info *tbl_info,
+			     unsigned int level, vaddr_t va_base, void *table)
+{
+	tbl_info->level = level;
+	tbl_info->table = table;
+	tbl_info->va_base = va_base;
+	tbl_info->shift = CORE_MMU_SHIFT_OF_LEVEL(level);
+	assert(level < RISCV_PGLEVELS);
+	tbl_info->num_entries = RISCV_PTES_PER_PT;
+}
+
+void core_mmu_get_entry_primitive(const void *table, size_t level,
+				  size_t idx, paddr_t *pa, uint32_t *attr)
+{
+	struct mmu_pgt *pgt = (struct mmu_pgt *)table;
+	struct mmu_pte *pte = core_mmu_table_get_entry(pgt, idx);
+
+	if (core_mmu_entry_is_valid(pte)) {
+		if (pa)
+			*pa = pte_to_pa(pte);
+		if (attr)
+			*attr = pte_to_mattr(level, pte);
+	}
+}
+
+void core_mmu_set_entry_primitive(void *table, size_t level, size_t idx,
+				  paddr_t pa, uint32_t attr)
+{
+	struct mmu_pgt *pgt = (struct mmu_pgt *)table;
+	struct mmu_pte *pte = core_mmu_table_get_entry(pgt, idx);
+	unsigned long perms = mattr_to_perms(level, attr)
+			    | PTE_V |  PTE_D | PTE_A;
+
+	core_mmu_entry_set(pte, core_mmu_pte_create(pa_to_ppn(pa), perms));
+}
+
+static void set_user_va_idx(struct mmu_partition *prtn)
+{
+	struct mmu_pgt *pgt;
+	struct mmu_pte *pte;
+	unsigned int idx;
+
+	pgt = core_mmu_get_base_pgt_va(prtn);
+
+	for (idx = 0 ; idx < RISCV_PTES_PER_PT; idx++) {
+		pte = core_mmu_table_get_entry(pgt, idx);
+		if (core_mmu_entry_is_invalid(pte)) {
+			user_va_idx = idx;
+			return;
+		}
+	}
+	assert(user_va_idx != -1);
+}
+
+static struct mmu_pte *core_mmu_get_user_mapping_entry
+			(struct mmu_partition *prtn)
+{
+	struct mmu_pgt *pgt = core_mmu_get_base_pgt_va(prtn);
+
+	assert(user_va_idx != -1);
+
+	return core_mmu_table_get_entry(pgt, user_va_idx);
+}
+
+void core_mmu_set_user_map(struct core_mmu_user_map *map)
+{
+	unsigned long satp = 0;
+	uint32_t exceptions = thread_mask_exceptions(THREAD_EXCP_ALL);
+	struct mmu_partition *prtn = core_mmu_get_prtn();
+	struct mmu_pte *pte;
+
+	satp = read_satp();
+	pte = core_mmu_get_user_mapping_entry(prtn);
+	if (map && map->user_map) {
+		core_mmu_entry_set(pte, core_mmu_ptp_create(pa_to_ppn
+				   ((paddr_t)map->user_map)));
+		core_mmu_table_write_barrier();
+		satp |= SHIFT_U64(map->asid, RISCV_SATP_ASID_SHIFT);
+		write_satp(satp);
+	} else {
+		core_mmu_entry_set(pte, 0);
+		core_mmu_table_write_barrier();
+	}
+
+	thread_unmask_exceptions(exceptions);
+}
+
+void core_mmu_get_user_va_range(vaddr_t *base, size_t *size)
+{
+	assert(user_va_idx != -1);
+
+	if (base)
+		*base =  (vaddr_t)1 << CORE_MMU_USER_TABLE_SHIFT;
+
+	if (size)
+		*size = (RISCV_PTES_PER_PT - 1) << CORE_MMU_USER_TABLE_SHIFT;
+}
+
+void core_mmu_get_user_pgdir(struct core_mmu_table_info *pgd_info)
+{
+	vaddr_t va_range_base;
+	struct mmu_partition *prtn = core_mmu_get_prtn();
+	struct mmu_pgt *pgt = core_mmu_get_ta_pgt_va(prtn);
+
+	core_mmu_get_user_va_range(&va_range_base, NULL);
+	/*
+	 * TODO: sync with core_mmu_set_info_table
```

Comment:
Please fix the TODO.

Reply:
This needs a generic way of iterating through page table levels. It depends on https://github.com/OP-TEE/optee_os/blob/master/core/mm/core_mmu.c#L1908 on RISC-V next level of level is (level - 1).

Reply:
OK, I think you'll have to accept that the software defined levels increase in the other direction. I mean there are only two directions to choose from and we are already using this.

Reply:
- For ARM, start from `CORE_MMU_BASE_TABLE_LEVEL` and increment, https://github.com/OP-TEE/optee_os/blob/master/core/mm/core_mmu.c#L1668 https://github.com/OP-TEE/optee_os/blob/master/core/mm/core_mmu.c#L1695
In our definition for RISC-V `CORE_MMU_BASE_TABLE_LEVEL` is LEVEL 2 in case using Sv39, LEVEL 1 in case of using Sv39 for RV32 and decrement level to go deeper.

- This assertion does not work also for RISC-V https://github.com/OP-TEE/optee_os/blob/master/core/mm/core_mmu.c#L1673 We set `CORE_MMU_PGDIR_LEVEL` to mega page, this is an intermediate level, we can map at giga page.